pyro/ops/stats.py

```diff
@@ -397,3 +397,44 @@ def fit_generalized_pareto(X):
     k = k * N / (N + 10.0) + 5.0 / (N + 10.0)
 
     return k, sigma
+
+
+def crps_empirical(pred, truth):
+    """
+    Computes negative Continuous Ranked Probability Score CRPS* [1] between a
+    set of samples ``pred`` and true data ``truth``. This uses an ``n log(n)``
+    time algorithm to compute a quantity equal that would naively have
+    complexity quadratic in the number of samples ``n``::
+
+        CRPS* = E|pred - truth| - 1/2 E|pred - pred'|
+              = (pred - truth).abs().mean(0)
+              - (pred - pred.unsqueeze(1)).abs().mean([0, 1]) / 2
+
+    Note that for a single sample this reduces to absolute error.
+
+    References
+    [1] `Strictly Proper Scoring Rules, Prediction, and Estimation`
+    Tilmann Gneiting, Adrian E. Raftery (2007)
+    https://www.stat.washington.edu/raftery/Research/PDF/Gneiting2007jasa.pdf
+
+    :param torch.Tensor pred: A set of sample predictions batched on rightmost dim.
+        This should have shape ``(num_samples,) + truth.shape``.
+    :param torch.Tensor truth: A true tensor.
+    :return: A tensor of shape ``truth.shape``.
+    :rtype: torch.Tensor
+    """
+    if pred.dim() != 1 + truth.dim() or pred.shape[1:] != truth.shape:
+        raise ValueError(f"Expected pred to have one extra sample dim on left. "
+                         f"Actual shapes: {pred.shape} versus {truth.shape}")
+    opts = dict(device=pred.device, dtype=pred.dtype)
+    num_samples = pred.size(0)
+    if num_samples == 1:
+        return (pred[0] - truth).abs()
+
+    pred = pred.sort(dim=0).values
```

Comment:
is the n log n algorithm from [hersbach](https://journals.ametsoc.org/doi/full/10.1175/1520-0434%282000%29015%3C0559%3ADOTCRP%3E2.0.CO%3B2)?

Reply:
Yeah, it looks like eqn. 20 in that reference contains the same trick. They sure spend lots of words to describe 10 lines of code 😓